walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,156 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> slots = new ArrayList<TimeRange>();
+
+    ArrayList<Event> eventsArr = sortEventsByStart(events);
+
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return slots;
+    }
+
+    slots.add(TimeRange.WHOLE_DAY);
+
+    if (!request.getOptionalAttendees().isEmpty()) {
+      ArrayList<TimeRange> optionalSlots = findTimeSlots(eventsArr, request, slots, true);
+      if (optionalSlots.isEmpty() && !request.getAttendees().isEmpty()) {
+        return findTimeSlots(eventsArr, request, slots, false);
+      }
+      return optionalSlots;
+    } else {
+      return findTimeSlots(eventsArr, request, slots, false);
+    }
+  }
+
+  /*
+   * Sort events by the start of their TimeRanges.
+   */
+  private ArrayList<Event> sortEventsByStart(Collection<Event> events) {
+    ArrayList<Event> eventsArr = new ArrayList<>(events);
+    Collections.sort(eventsArr, new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return Long.compare(a.getWhen().start(), b.getWhen().start());
+      }
+    });
+    return eventsArr;
+  }
+
+  /*
+   * Returns a list with all the available time slots when the requested meeting
+   * can take place.
+   */
+  private ArrayList<TimeRange> findTimeSlots(Collection<Event> events, MeetingRequest request,
+    ArrayList<TimeRange> slots, boolean hasOptionalAttendees) {
+
+    // A copy of slots that stores new changes to slots without interferring
+    // with the for loop below.
+    ArrayList<TimeRange> curSlots = new ArrayList<>(slots);
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      slots = curSlots;
+
+      // Determine whether event's attendees overlaps with request's attendees.
+      if (hasOptionalAttendees && !hasAllAttendees(event, request)) {
+        continue;
+      }
+      if (!hasOptionalAttendees && !hasRequiredAttendees(event, request)) {
+        continue;
+      }
```

Comment:
So here you're trying to determine if this even should "disqualify" the TimeRange that this event takes place during.  If either of these checks are true, the event is not a problem and shouldn't be considered.  Otherwise you have to figure out how to carve out the event's time slot from your running list of options.

I think this can be documented better via comments and also variable/method names since I found it to be a little difficult to follow this code.